flow/probes/dpdk.go

```diff
@@ -0,0 +1,196 @@
+
+	for i := uint(0); i < nbPackets; i++ {
+		packet := gopacket.NewPacket(packets[i].GetRawPacketBytes(), layers.LayerTypeEthernet, gopacket.Default)
+		if flowPackets := flow.PacketsFromGoPacket(&packet, 0, -1, nil); len(flowPackets.Packets) > 0 {
+			ctx <- flowPackets
+		}
+		next[i] = false
+	}
+}
+
+func l3Splitter(currentPacket *packet.Packet, context dpdkflow.UserContext) uint {
+	ipv4, ipv6 := currentPacket.ParseAllKnownL3()
+	if ipv4 != nil {
+		h := (ipv4.SrcAddr>>24)&0xff ^ (ipv4.DstAddr>>24)&0xff ^
+			(ipv4.SrcAddr>>16)&0xff ^ (ipv4.DstAddr>>16)&0xff ^
+			(ipv4.SrcAddr>>8)&0xff ^ (ipv4.DstAddr>>8)&0xff ^
+			(ipv4.SrcAddr)&0xff ^ (ipv4.DstAddr)&0xff
+		return uint(h) % dpdkNBWorkers
+	}
+	if ipv6 != nil {
+		h := uint(0)
+		for i := range ipv6.SrcAddr {
+			h = h ^ uint(ipv6.SrcAddr[i]^ipv6.DstAddr[i])
+		}
+		return h % dpdkNBWorkers
+	}
+	return 0
+}
+
+type myctx chan *flow.Packets
+
+func (c myctx) Copy() interface{} {
+	return c
+}
+
+func getDPDKMacAddress(port int) string {
+	mac := dpdkflow.GetPortMACAddress(uint8(port))
+	macAddr := ""
+	for i, m := range mac {
+		macAddr += fmt.Sprintf("%02x", m)
+		if i < (len(mac) - 1) {
+			macAddr += ":"
+		}
+	}
+	return macAddr
+}
+
+// NewDPDKProbesHandler creates a new gopacket probe in the graph
+func NewDPDKProbesHandler(g *graph.Graph, fta *flow.TableAllocator, fcpool *analyzer.FlowClientPool) (*DPDKProbesHandler, error) {
+	ports := config.GetConfig().GetStringSlice("dpdk.ports")
+	nbWorkers := config.GetConfig().GetInt("dpdk.workers")
+
+	nbPorts := len(ports)
+	if nbWorkers == 0 || nbPorts == 0 {
+		return nil, fmt.Errorf("DPDK flow porbe is not configured")
+	}
+	dpdkNBWorkers = uint(nbWorkers)
+
+	cfg := &dpdkflow.Config{
+		LogType: dpdkcommon.Initialization,
+	}
+	debug := config.GetConfig().GetInt("dpdk.debug")
+	if debug > 0 {
+		cfg.LogType = dpdkcommon.Debug
+		cfg.DebugTime = uint(debug * 1000)
+	}
+	dpdkflow.SystemInit(cfg)
+
+	opts := flow.TableOpts{
+		RawPacketLimit: 0,
+		TCPMetric:      false,
+		SocketInfo:     false,
+	}
+
+	dph := &DPDKProbesHandler{
+		graph:          g,
+		fta:            fta,
+		flowClientPool: fcpool,
+	}
+
+	for _, port := range ports {
+		inport, err := strconv.Atoi(port)
+		if err != nil {
+			return nil, fmt.Errorf("misconfiguration of DPDK port %v", port)
+		}
+		m := graph.Metadata{
+			"Name":      fmt.Sprintf("DPDK-port-%d", inport),
+			"EncapType": "ether",
+			"IfIndex":   inport,
+			"MAC":       getDPDKMacAddress(inport),
+			"Driver":    "dpdk",
+			"State":     "UP",
+		}
+		dpdkNode := g.NewNode(graph.GenID(), m)
+		hostNode := g.LookupFirstNode(graph.Metadata{
+			"Name": g.GetHost(),
+			"Type": "host",
+		})
+		topology.AddOwnershipLink(g, hostNode, dpdkNode, nil)
+
+		inputFlow := dpdkflow.SetReceiver(uint8(inport))
+		outputFlows := dpdkflow.SetSplitter(inputFlow, l3Splitter, uint(dpdkNBWorkers), nil)
+
+		for i := 0; i < nbWorkers; i++ {
+			ft := fta.Alloc(dph.asyncFlowPipeline, string(dpdkNode.ID), opts)
```

Comment:
not dpdkNode.ID but dpdkNode.GetFieldString("TID")

Reply:
doen